### PR TITLE
[Merged by Bors] - feat(ContextFreeGrammar): More lemmas about reversal

### PR DESCRIPTION
**Pull request by YaelDillies**
A lot of API was missing

---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 1]
```diff
@@ -200,59 +205,98 @@ proof_wanted Language.isContextFree_iff {L : Language T} :
...
+
+@[simp] lemma language_reverse : g.reverse.language = g.language.reverse := by ext; simp
+
+end ContextFreeGrammar
 
 /-- The class of context-free languages is closed under reversal. -/
-theorem Language.IsContextFree.reverse (L : Language T) :
-    L.IsContextFree → L.reverse.IsContextFree :=
-  fun ⟨g, hg⟩ => ⟨g.reverse, hg ▸ Set.ext g.mem_reverse_language_iff_reverse_mem_language⟩
+theorem Language.IsContextFree.reverse [DecidableEq T] (L : Language T) :
```
Comment (YaelDillies) [inline Mathlib/Computability/ContextFreeGrammar.lean:298]:
```suggestion
theorem Language.IsContextFree.reverse (L : Language T) :
```

**Comment by madvorak** [comment]
I like the new API but I don't like the new `variable`s cluttering my proofs that didn't need those variables. Can we introduce them right before they are needed in each section?

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 2]
```diff
@@ -200,59 +205,98 @@ proof_wanted Language.isContextFree_iff {L : Language T} :
...
+lemma rewrites_reverse : r.reverse.Rewrites u.reverse v.reverse ↔ r.Rewrites u v :=
+  ⟨fun h ↦ by simpa using h.reverse, .reverse⟩
+
+@[simp] lemma rewrites_reverse_comm : r.reverse.Rewrites u v ↔ r.Rewrites u.reverse v.reverse := by
+  rw [← rewrites_reverse, reverse_reverse]
+
+end ContextFreeRule
+
+namespace ContextFreeGrammar
+variable {g : ContextFreeGrammar T} {s u v : List (Symbol T g.NT)} {w : List T}
```
Comment (madvorak) [inline Mathlib/Computability/ContextFreeGrammar.lean:244]: remove `s`

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 3]
```diff
@@ -200,59 +205,98 @@ proof_wanted Language.isContextFree_iff {L : Language T} :
...
+@[simp] lemma language_reverse : g.reverse.language = g.language.reverse := by ext; simp
+
+end ContextFreeGrammar
 
 /-- The class of context-free languages is closed under reversal. -/
 theorem Language.IsContextFree.reverse (L : Language T) :
-    L.IsContextFree → L.reverse.IsContextFree :=
-  fun ⟨g, hg⟩ => ⟨g.reverse, hg ▸ Set.ext g.mem_reverse_language_iff_reverse_mem_language⟩
+    L.IsContextFree → L.reverse.IsContextFree := by
+  rintro ⟨g, rfl⟩; exact ⟨g.reverse, by simp⟩
```
Comment (madvorak) [inline Mathlib/Computability/ContextFreeGrammar.lean:300]: Can we remove the `;` and make it two lines please? Everything else in this file is written "one tactic a line".

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 4]
```diff
@@ -19,8 +19,11 @@ nonterminal symbol on the left-hand side of each rule.
 * `Language.IsContextFree.reverse`: The class of context-free languages is closed under reversal.
 -/
 
+open Function
```
Comment (madvorak) [inline Mathlib/Computability/ContextFreeGrammar.lean:22]: This is afaik used only in one section. Can we move it there please?

Comment (YaelDillies) [reply Mathlib/Computability/ContextFreeGrammar.lean:22]: It's used in two sections, actually

File: Mathlib/Computability/ContextFreeGrammar.lean (hunk 3 above)
Comment (YaelDillies) [reply Mathlib/Computability/ContextFreeGrammar.lean:300]: https://github.com/leanprover-community/mathlib4/blob/32419692d138999119817d8612c02c254d26fcba/Mathlib/Computability/ContextFreeGrammar.lean#L80

would like to disagree!

File: Mathlib/Computability/ContextFreeGrammar.lean [hunk 5]
```diff
@@ -200,59 +205,97 @@ proof_wanted Language.isContextFree_iff {L : Language T} :
...
+
+protected lemma Derives.reverse (hg : g.Derives u v) : g.reverse.Derives u.reverse v.reverse := by
+  induction hg with
+  | refl => rfl
+  | tail _ orig ih => exact ih.trans_produces orig.reverse
+
+lemma derives_reverse : g.reverse.Derives u.reverse v.reverse ↔ g.Derives u v :=
+  ⟨fun h ↦ by convert h.reverse <;> simp, .reverse⟩
+
+@[simp] lemma derivees_reverse_comm : g.reverse.Derives u v ↔ g.Derives u.reverse v.reverse := by
```
Comment (TwoFX) [inline Mathlib/Computability/ContextFreeGrammar.lean:283]:
```suggestion
@[simp] lemma derives_reverse_comm : g.reverse.Derives u v ↔ g.Derives u.reverse v.reverse := by
```

File: Mathlib/Computability/ContextFreeGrammar.lean (hunk 3 above)
Comment (madvorak) [reply Mathlib/Computability/ContextFreeGrammar.lean]: Oooops. It looks like a refactor somebody did after me.

File: Mathlib/Computability/ContextFreeGrammar.lean (hunk 4 above)
Comment (madvorak) [reply Mathlib/Computability/ContextFreeGrammar.lean:22]: My bad!

File: Mathlib/Computability/ContextFreeGrammar.lean (hunk 5 above)
Comment (madvorak) [reply Mathlib/Computability/ContextFreeGrammar.lean:283]: I thought I already fixed this typo. Apparently, it was on the other branch.

File: Mathlib/Computability/ContextFreeGrammar.lean (hunk 3 above)
Comment (YaelDillies) [reply Mathlib/Computability/ContextFreeGrammar.lean]: Apparently it comes straight from #6385 ! Maybe past Martin would disapprove current Martin 😆

**Review by TwoFX — APPROVED** [review]
Thanks!
bors r+

**Comment by TwoFX** [comment]
bors d+

**Comment by YaelDillies** [comment]
bors merge

**Pull request closed without merging**